### ISSUE TITLE
Fully customizable pane and window layouts

### ISSUE THREAD

**Issue by jmcphers** [rstudio/rstudio#2879]
Today, RStudio's pane layout always deals in quadrants. Two quadrants are occupied by the Source and Console panes, and a second two by sets of tabs. The user can choose which panes appear in which quadrants.

Most IDEs have a somewhat more flexible system which allows for arbitrary tab arrangement, often using mouse gestures to edit configuration on the fly -- docking to sides, splitting and tabbing, popping out to floating windows, etc. In the last few years several JavaScript libraries have been released which make this relatively straightforward to implement (at least from a UX standpoint). For example:

https://phosphorjs.github.io/ - used by Jupyter

http://www.dockspawn.com/ - with some design inspiration from Visual Studio

http://golden-layout.com/ - emphasizes pleasing layouts & multi-screen

Now that we're on GWT 2.8, we can use [JsInterop](http://www.gwtproject.org/doc/latest/DevGuideCodingBasicsJsInterop.html), which will make it much more straightforward to blend our own tab internals with an external library.

**Comment by philiporlando** [comment]
I really want to view the terminal and R console simultaneously. It would be awesome if we could move the terminal into the plotting or environment panes. 

![rstudioterminal](https://user-images.githubusercontent.com/24445174/45913035-3c7dc700-bde0-11e8-9a77-19beadb71974.png)

**Comment by dracodoc** [comment]
Fully customized layout would be great, though it could be big project.

For now I'm bothered by one little annoyance: I cannot turn off the pane I don't need, like `connection` pane. I don't need it and I have limited screen real estate, the `connection` pane make `build` pane invisible for my favored pane sizes.

**Comment by moodymudskipper** [comment]
What I usually do know is :

* detach my current script window and put it full screen on a vertical screen.
* make a plot window and put it on another screen.
* have the main studio window on a 3rd one

It's frustrating though when I need to switch editing scripts. Being able to split the 4 parts completely, or combine them as I please would be fantastic

**Comment by vikram-rawat** [comment]
Just to add on. 

If I can see my console output, plot output and viewer output on a same pane. That way I can always make it full size and see everything and it will switch between console and plot output whenever I give one.

Trust me it's easier to develop shiny apps that way when both of your output are always fill screen.

**Comment by BrainStormCenter** [comment]
In addition to splitting panes, the ability to move tabs from one pane to another would be a very nice addition. Does anyone know the current possibilities of these requests being incorporated
in a future release?

**Comment by gtritchie** [comment]
Tagging as an accessibility issue. Screen-reader users may find a single pane layout much easier to navigate (one pane, all the tabs in that pane, including ability to get rid of tabs the user never uses).

**Comment by schradj** [comment]
Would like to give a bump to the specific feature of splitting source panes to view multiple locations in a single file simultaneously. a la gvim (and probably other editors)

**Comment by picousse** [comment]
Hi,
Not sure it fits here.
Big fan of the extra source pane in the beta.
Unfortunately, I have to go to global settings, Pane layout, add column to get it up and working.

And It doesn't seem to store that.

A shortcut/button to create new pane would be nice. There is already a button to get a source pane popup, an extra one to create a new pane would be nice as well.

Keep up the good work! this extra pane is simple awesome.

**Comment by jucor** [comment]
Having vertical splits to see a same file in two different places side-by-side is such a time-saving feature. For example: having the definition of a function and the place you call it, both visible at the same time. With 80-col style and wide screens, it fits perfectly.
Having this feature in RStudio would be a relief: I very much miss it!

**Comment by jucor** [comment]
Please don't stale. Clearly need the capacity to edit two parts of a same
file in two panes.

**Comment by slager** [comment]
+1

**Comment by pat-hearps** [comment]
Have just tried the extra source panes in v1.4, is very good. Would be better if it's not just tacked onto the left side, but could be in essence a split version of the main source panel, with the console/terminal underneath both of them.

**Comment by doncqueurs** [comment]
I would appreciate a three column layout in the RStudio IDE. I have one of those wide monitors (5200px). Having a full height column (pane) for both the editor and the console would be great and then the enviornments/GIT/etc and Files/Plots/Packages/Help/View/etc. can be in the third column sharing the height.

**Comment by ElinaMX** [comment]
As said @natereal, it would be great to be able to set a new source pane in the middle of the window as the main script, and keep the consol/terminal + an other source pane (for yaml, css, notes, or usefull chunks for example) on the left while having the classic environment + files/plots on the right.

**Comment by marketresearchru** [comment]
In light of the Posit Assistant planned addition, this feature becomes crucial.
